File: Data Structures & Algorithms/valid-sudoku/submission-0.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        int[] rows = new int[9];
        int[] cols = new int[9];
        int[] squares = new int[9];

        for (int r = 0; r < 9; r++) {
            for (int c = 0; c < 9; c++) {

                // Non-Word
                if (board[r][c] == '.') continue;

                // Get value
                int val = board[r][c] - '1';
                
                // Check if val exists in row
                // Check if val exists in col
                // Checks if val already exists in corresponding 3x3 square
                // Computes index (0-8) of the square the cell (r, c) belongs to
                if ((rows[r] & (1 << val)) > 0 || 
                    (cols[c] & (1 << val)) > 0 || 
                    (squares[(r / 3) * 3 + (c / 3)] & (1 << val)) > 0) {
                    return false;
                }

               
                rows[r] |= (1 << val);  // Add to Row    
                cols[c] |= (1 << val);  // Add to Cols
                squares[(r / 3) * 3 + (c / 3)] |= (1 << val); // Add to Squares
            }
        }
        return true;
    }
}

File: Data Structures & Algorithms/search-for-word/submission-0.py
"""
CAT

board = [
  ["A","B","C","D"],
  ["S","A","A","T"],
  ["A","C","A","E"]
]

word[i], i = 0, C

bfs(i, position, curr)

"""

class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        directions = [[-1, 0], [1, 0], [0, -1], [0, 1]]
        ROWS, COLS = len(board), len(board[0])
        visited = set()

        def dfs(r, c, i):
            if i == len(word):
                return True
            if (r < 0 or c < 0 or r >= ROWS or c >= COLS or 
                board[r][c] != word[i] or (r, c) in visited):
                return False

            visited.add((r, c))
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                if dfs(nr, nc, i + 1):
                    return True
            visited.remove((r, c))
            return False

        for row in range(ROWS):
            for col in range(COLS):
                if dfs(row, col, 0):
                    return True
        return False